[ICDS Dashboard]: AWW Activity Report

ShubhamBansal1997 opened the pull request:
Sample output files
[Aww_Activity_Report__06-06-2020__20-41-11.zip](https://github.com/dimagi/commcare-hq/files/4740504/Aww_Activity_Report__06-06-2020__20-41-11.zip)
[Aww_Activity_Report__06-06-2020__20-40-26.xlsx](https://github.com/dimagi/commcare-hq/files/4740505/Aww_Activity_Report__06-06-2020__20-40-26.xlsx)


<!--- Provide a link to the ticket or document which prompted this change -->

##### SUMMARY
https://app.asana.com/0/1175081284767805/1171569920995114/f
<!--- Describe the change below, including rationale and design decisions -->

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### RISK ASSESSMENT / QA PLAN
<!-- Does this need QA before or after merge? Link QA ticket. -->

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

ShubhamBansal1997 committed:
🎉 Intial Commit for the AWW Activity Report

ShubhamBansal1997 committed:
✨ updated js for frontend and backend for ordering

ShubhamBansal1997 committed:
🔀 Merge branch 'master' into sb/aww_activity_report

ShubhamBansal1997 committed:
✨ added the excel function
✨ added the frontend code
✅ added the tests

ShubhamBansal1997 committed:
✅ added frontend test
move the report to feature flag

czue commented:
planning to let @Rohit25negi handle review, though as always let me know if you'd like me to take a pass

ShubhamBansal1997 committed:
✅ fixed frontend tests

Rohit25negi commented on custom/icds_reports/sqldata/exports/aww_activity_report.py:
```diff
@@ -0,0 +1,80 @@
+
+        filters = {}
+
+        if self.loc_level == 4:
```
this will result in filter dict like 
```
{'<location>_id':<location_id>}

``` 

yeah?

I think filter should also include all the parent locations as well. It would improve the cost of the the query. For example:
if you will query for filters = {block_id:23456678}. It will look for that block in all the states. So good to have all the ancestors in the filter.

ShubhamBansal1997 replied:
updated

Rohit25negi commented on custom/icds_reports/static/icds_reports/js/spec/download.directive.spec.js:
```diff
@@ -6,7 +6,7 @@ var pageData = hqImport('hqwebapp/js/initial_page_data');
 describe('Download Directive', function () {
 
-    var numberOfReports = 12;
+    var numberOfReports = 13;
```
why did this change?

ShubhamBansal1997 replied:
reverted

Rohit25negi commented on custom/icds_reports/migrations/0193_add_fields_aww_inactivity_report.py:
```diff
@@ -0,0 +1,25 @@
+            name='no_of_days_since_start',
+            field=models.PositiveIntegerField(blank=True, null=True),
+        ),
+    ]
```
it would also be helpful to check if this table have the required indexes present. If not, we should create them with this migration. It will make the query faster.

Rohit25negi reviewed (changes requested):
left a few change request and a few questions.

ShubhamBansal1997 commented on custom/icds_reports/static/js/directives/download/download.directive.js:
```diff
@@ -468,7 +470,8 @@ function DownloadController($scope, $rootScope, $location, locationHierarchy, lo
             ladySupervisorReportErrors = vm.isLadySupervisorSelected() && !vm.isStateSelected();
         }
-        return beneficiaryListErrors || incentiveReportErrors || ladySupervisorReportErrors || growthListErrors;
+        var awwActvityReportErrors = vm.isAwwActivityReportSelected() && !vm.isStateSelected();
```
Yes, it is meant to be at state level and below till sector level

ShubhamBansal1997 committed:
👌 removed the print statements
✅ fixed frontend tests

ShubhamBansal1997 committed:
👌 fixed as per reivew
🗃 added indexes
🔧 fixed the aggergation relared function
✨ improved the query function

ShubhamBansal1997 committed:
Merge branch 'master' into sb/aww_activity_report

ShubhamBansal1997 committed:
🔧 fixed migration issues

ShubhamBansal1997 committed:
✏️ fixing the minor typo in sql statement

ShubhamBansal1997 committed:
🔧 fixed helper function for inactive awws

ShubhamBansal1997 committed:
🐛 correcting the date_part function

ShubhamBansal1997 committed:
🐛 correcting the date_part function

ShubhamBansal1997 committed:
🔧 added the _aggregate_inactive_aww_agg to agg

ShubhamBansal1997 committed:
✅ improved tests cases

ShubhamBansal1997 committed:
🚨 fix lynt issues

ShubhamBansal1997 committed:
✅ fixed tests and tasks.py

ShubhamBansal1997 committed:
✅ fixed test cases

ShubhamBansal1997 committed:
✅ fixed tests

ShubhamBansal1997 commented on custom/icds_reports/utils/aggregation_helpers/distributed/inactive_awws.py:
```diff
@@ -91,7 +91,15 @@ def aggregate_query(self):
+                    ELSE NULL END,
+                no_of_days_inactive = CASE
+                    WHEN GREATEST(agg_table.last_submission, ut.last_submission) IS DISTINCT FROM NULL
+                    THEN DATE_PART('day', NOW() - GREATEST(agg_table.last_submission, ut.last_submission))
```
updated

ShubhamBansal1997 commented on custom/icds_reports/tasks.py:
```diff
@@ -1029,9 +1033,39 @@ def prepare_excel_reports(config, aggregation_level, include_test, beta, locatio
         data_type = 'Child Growth Tracker Report__{}'.format(formatted_timestamp)
+    elif indicator == AWW_ACTIVITY_REPORT:
+        data_type = 'AWW_Activity_Report'
+        print(config)
```
updated

ShubhamBansal1997 committed:
🚨 removed print statements

ShubhamBansal1997 committed:
🐛 added missing variable to class constructor

ShubhamBansal1997 committed:
✅ fixed tests

ShubhamBansal1997 committed:
✅ fixed tests

ShubhamBansal1997 commented:
@Rohit25negi Please re-review

Rohit25negi commented on custom/icds_reports/migrations/0195_add_fields_aww_inactivity_report.py:
```diff
@@ -0,0 +1,28 @@
+            field=models.PositiveIntegerField(blank=True, null=True),
+        ),
+        migrations.RunSQL(
+            f"CREATE INDEX idx_aww_activity_state_district_block_sector ON {AggregateInactiveAWW._meta.db_table} (state_id, district_id, block_id, supervisor_id)")
```
did you check if table already have indexes?

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/distributed/inactive_awws.py:
```diff
@@ -107,4 +115,5 @@ def aggregate_query(self):
             table_name=self.aggregate_parent_table,
             ucr_table_query=ucr_query,
             awc_location_table_name='awc_location_local',
+            now=datetime.datetime.utcnow()
```
you can directly pass date here instead of passing time and converting it to date in the query.

Rohit25negi commented on custom/icds_reports/tests/agg_tests/test_aggregation_script.py:
```diff
@@ -415,15 +415,15 @@ def setUpClass(cls):
             cursor.execute(missing_location_query)
         records = AggregateInactiveAWW.objects.filter(first_submission__isnull=False)
-        self.assertEqual(records.count(), 0)
+        self.assertEqual(records.count(), 46)
```
This change invalidates the significance of this test case which is to test just the missing location query in which there should be no rows with first_submission is not null
Better would be to call the aggregate_aww function in the `test_aww_activity_export` function below before you pull the data.

Rohit25negi reviewed (changes requested):
looks good overall. Left one comment and one question should be easy to address

ShubhamBansal1997 committed:
👌changed the datetime to date
reverted the correct tests

ShubhamBansal1997 committed:
✅ fixed tests

ShubhamBansal1997 committed:
Merge branch 'master' into sb/aww_activity_report

ShubhamBansal1997 commented:
@Rohit25negi please re-review

Rohit25negi reviewed (approved):
looks good. 
It may just be a preferential thing, It makes the review a lot harder if we just reformat the file, it is hard to identify what snippet is actually the change and what snippet is just formatting diff.

ShubhamBansal1997 commented:
https://travis-ci.org/github/dimagi/commcare-hq/builds/696035296 Build passed
